### PR TITLE
Generate valid individual identification numbers kk_KZ

### PR DESCRIPTION
**Pull request by YerlenZhubangaliyev**
'individualIdentificationNumber' method now generates valid individual identification numbers

### CODE REVIEW

File: src/Faker/Provider/kk_KZ/Person.php [hunk 1]
```diff
@@ -134,22 +177,68 @@ class Person extends \Faker\Provider\Person
...
+     */
+    private static function getCenturyByYear($year)
+    {
+        if ($year >= 1800 && $year <= 1899) {
+            return self::CENTURY_19TH;
+        } elseif ($year >= 1900 && $year <= 1999) {
+            return self::CENTURY_20TH;
+        } elseif ($year >= 2000 && $year <= DateTime::year()) {
+            return self::CENTURY_21ST;
+        }
```
Comment (ankitpokhrel) [inline src/Faker/Provider/kk_KZ/Person.php]: You can flip conditions and remove `&&` 
```
if ($year >= 2000 && $year <= DateTime::year()) {
    return self::CENTURY_21ST;
} elseif ($year >= 1900) {
    return self::CENTURY_20TH;
} elseif ($year >= 1800) {
    return self::CENTURY_19TH;
}
```

File: src/Faker/Provider/kk_KZ/Person.php [hunk 2]
```diff
@@ -134,22 +177,68 @@ class Person extends \Faker\Provider\Person
     );
 
     /**
+     * @param  integer $year
+     *
+     * @return integer
```
Comment (ankitpokhrel) [inline src/Faker/Provider/kk_KZ/Person.php]: `return integer|null`

File: src/Faker/Provider/kk_KZ/Person.php [hunk 3]
```diff
@@ -134,22 +177,68 @@ class Person extends \Faker\Provider\Person
...
         if (!$birthDate) {
-            $birthDate = \Faker\Provider\DateTime::dateTimeBetween();
+            $birthDate = DateTime::dateTimeBetween();
         }
 
-        $dateAsString       = $birthDate->format('ymd');
-        $genderAndCenturyId = (string) static::numberBetween(1, 6);
-        $randomDigits       = (string) static::numerify('#####');
+        $dateAsString     = $birthDate->format('ymd');
+        $totalPopulation  = rand(1000, 2000);
```
Comment (fzaninotto) [inline src/Faker/Provider/kk_KZ/Person.php]: never, ever use `rand` in Faker. See explanation in https://github.com/fzaninotto/Faker/blob/master/CONTRIBUTING.md

File: test/Faker/Provider/kk_KZ/PersonTest.php [hunk 4]
```diff
@@ -19,13 +20,28 @@ public function setUp()
...
+        $birthDate                      = DateTime::dateTimeBetween('-30 years', '-10 years');
         $individualIdentificationNumber = $this->faker->individualIdentificationNumber($birthDate);
-        $birthDateAsString              = $birthDate->format('ymd');
+        $sum                            = 0;
 
-        $this->assertRegExp(
-            "/^(" . $birthDateAsString . ")([1-6]{1})(\\d{5})$/",
-            $individualIdentificationNumber
-        );
+        for ($i = 0; $i <= 10; $i++) {
```
Comment (fzaninotto) [inline test/Faker/Provider/kk_KZ/PersonTest.php]: don't duplicate the checksum code here. Instead, refactor it into a utility function, and use it both in the generator and in the test.

File: src/Faker/Provider/kk_KZ/Person.php (hunk 3 above)
Comment (YerlenZhubangaliyev) [inline src/Faker/Provider/kk_KZ/Person.php]: done

File: test/Faker/Provider/kk_KZ/PersonTest.php (hunk 4 above)
Comment (YerlenZhubangaliyev) [inline test/Faker/Provider/kk_KZ/PersonTest.php]: done

File: src/Faker/Provider/kk_KZ/Person.php [hunk 5]
```diff
@@ -134,22 +177,80 @@ class Person extends \Faker\Provider\Person
...
      * @link   http://egov.kz/wps/portal/Content?contentPath=%2Fegovcontent%2Fcitizen_migration%2Fpassport_id_card%2Farticle%2Fiin_info&lang=en
+     * @link   https://ru.wikipedia.org/wiki/%D0%98%D0%BD%D0%B4%D0%B8%D0%B2%D0%B8%D0%B4%D1%83%D0%B0%D0%BB%D1%8C%D0%BD%D1%8B%D0%B9_%D0%B8%D0%B4%D0%B5%D0%BD%D1%82%D0%B8%D1%84%D0%B8%D0%BA%D0%B0%D1%86%D0%B8%D0%BE%D0%BD%D0%BD%D1%8B%D0%B9_%D0%BD%D0%BE%D0%BC%D0%B5%D1%80
+     *
      * @param  \DateTime $birthDate
+     * @param  integer   $gender
+     *
      * @return string 12 digits, like 780322300455
      */
-    public static function individualIdentificationNumber(\DateTime $birthDate = null)
+    public static function individualIdentificationNumber(\DateTime $birthDate = null, $gender = self::GENDER_MALE)
```
Comment (ankitpokhrel) [inline src/Faker/Provider/kk_KZ/Person.php]: Haven't tested, but looks like your methods can be refactored like this:

```php
public static function individualIdentificationNumber(\DateTime $birthDate = null, $gender = self::GENDER_MALE)
{
    if (!$birthDate) {
        $birthDate = DateTime::dateTimeBetween();
    }

    $population = mt_rand(1000, 2000);
    $century    = self::getCenturyByYear((int) $birthDate->format('Y'));
    
    $iin  = $birthDate->format('ymd');
    $iin .= (string) self::$genderCenturyMap[$gender][$century];
    $iin .= (string) $population;

    return  $iin . (string) self::checkSum($iin);
}

/**
 * @param string $iinValue
 *
 * @return integer
 */
public static function checkSum($iinValue)
{
    $controlDigit = self::getControlDigit($iinValue, self::$firstSequenceBitWeights);

    if ($controlDigit === 10) {
        return self::getControlDigit($iinValue, self::$secondSequenceBitWeights);
    }

    return $controlDigit;
}

/**
 * @param string $iinValue
 * @param array $sequence
 *
 * @return integer
 */
protected static function getControlDigit($iinValue, $sequence) 
{
    $sum = 0;

    for ($i = 0; $i <= 10; $i++) {
        $sum += (int) $iinValue[$i] * $sequence[$i];
    }

    return $sum % 11;
}
```

Comment (YerlenZhubangaliyev) [inline src/Faker/Provider/kk_KZ/Person.php]: done

Comment (ankitpokhrel) [inline src/Faker/Provider/kk_KZ/Person.php]: 👍

**Comment by fzaninotto** [comment]
Thanks!

**Comment by YerlenZhubangaliyev** [comment]
Thank you very much!

**Pull request merged**